main/tests/cypress/cypress/integration/project/grid/misc/expressions.spec.js

```diff
@@ -0,0 +1,306 @@
+			'Syntax error reading source'
+		)
+	})
+
+	it('Test a Grel language error', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		typeExpression('value.thisGrelFunctionDoesNotExists()')
+		cy.get('.expression-preview-parsing-status').contains(
+			'Unknown function thisGrelFunctionDoesNotExists'
+		)
+	})
+
+	it('Test a Python language error', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		cy.get('select[bind="expressionPreviewLanguageSelect"]').select(
+			'jython'
+		)
+		typeExpression('return value.thisPythonFunctionDoesNotExists()')
+
+		cy.get(
+			'.expression-preview-table-wrapper tr:nth-child(2) td:last-child'
+		).should('to.contain', 'Error: Traceback')
+	})
+
+	it('Test a Clojure language error', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		cy.get('select[bind="expressionPreviewLanguageSelect"]').select(
+			'clojure'
+		)
+		typeExpression('(.. value (thisClojureFunctionDoesNotExists) )')
+		cy.get(
+			'.expression-preview-table-wrapper tr:nth-child(2) td:last-child'
+		).should('to.contain', 'Error: No matching method')
+	})
+
+	it('Test switching from one langage to another', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		typeExpression('(.. value (toLowerCase) )')
+		// error is expected, this is clojure language
+		cy.get('.expression-preview-parsing-status').should(
+			'to.contain',
+			'Parsing error'
+		)
+		// switching to clojure
+		cy.get('select[bind="expressionPreviewLanguageSelect"]').select(
+			'clojure'
+		)
+		cy.get('.expression-preview-parsing-status').should(
+			'not.to.contain',
+			'Parsing error'
+		)
+	})
+
+	it('Test the preview (GREL)', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		typeExpression('value.toLowercase()')
+
+		cy.get(
+			'.expression-preview-table-wrapper tr:nth-child(1) td:last-child'
+		).should('to.contain', 'value.toLowercase()')
+		cy.get(
+			'.expression-preview-table-wrapper tr:nth-child(2) td:last-child'
+		).should('to.contain', 'butter,with salt')
+		cy.get(
+			'.expression-preview-table-wrapper tr:nth-child(3) td:last-child'
+		).should('to.contain', 'butter,whipped,with salt')
+	})
+
+	it('Test the help tab', function () {
+		cy.loadAndVisitProject('food.mini')
+		loadExpressionPanel()
+		cy.get('#expression-preview-tabs li').contains('Help').click()
+		cy.get('#expression-preview-tabs-help').should('be.visible')
+		cy.get('#expression-preview-tabs-help').should(
+			'to.contain',
+			`The current cell. It has a few fields: 'value', 'recon' and 'errorMessage'.`
+		)
+	})
+
+	it('Test the history behavior, ensure expressions are stored', function () {
+		cy.loadAndVisitProject('food.mini')
+		// Because history is shared across projects, we need to use an expression that is unique
+
+		// Use a first unique expression
+		const uniqueExpression = 'value.testing_One_' + Date.now() + '()'
```

Comment:
Because your current handling is not using a valid GREL expression (and we need to use valid GREL expressions , unless we are ERROR testing, for Star and History testing, because invalid GREL expressions won't make it into History and thus won't even show for you to Star them).

See how your current uniqueExpression value actually even blows up Cypress during the testing because it's an invalid GREL function for a non-error checking test.  Ensure to use only valid GREL functions for success testing:
![image](https://user-images.githubusercontent.com/986438/106291941-ddcb8200-6211-11eb-9a53-ac21ddd09158.png)

Reply:
@thadguidry see the comment above in the code
The thing is everything we enter as expressions is stored, across all projects.
So we can't know for sure if the expression found in the history comes from the tests of the previous run
Expressions needs to be unique, to guarantee idempotency

Reply:
@thadguidry invalid grel expressions are making it into history
See my local history
![Screen Shot 2021-01-29 at 16 35 42](https://user-images.githubusercontent.com/6267288/106294909-0ffbd480-6250-11eb-880a-f4b8ec61bf6d.png)

No, you have an issue with something else, and I've got the same but though I resolved it. Basically OpenRefine has updated the grid at some point, and Cypress is not able to query because the grid has been updated between 2 cypress statements about the same DOM element

What happens if you retry the test many times ? Does it pass at some point ?

Reply:
Oh wow!  Even invalid GREL expressions are being stored in history?  I did not even know that.
Hmm, I'd say that's probably not ideal or even useful at all for our users!
@wetneb Bug?